include/alpaka/queue/IGenericQueue.hpp

```diff
@@ -0,0 +1,57 @@
+/* Copyright 2019 Axel Huebl, Benjamin Worpitz, Matthias Werner
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/BoostPredef.hpp>
+
+// #include <alpaka/event/EventGeneric.hpp>
+// #include <alpaka/dev/DevCpu.hpp>
+
+namespace alpaka
+{
+    namespace event
+    {
+        template<typename TDev>
+        class EventGeneric;
+    }
+}
+
+namespace alpaka
+{
+    namespace queue
+    {
+#if BOOST_COMP_CLANG
+    // avoid diagnostic warning: "has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]"
+    // https://stackoverflow.com/a/29288300
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wweak-vtables"
+#endif
+
+        //#############################################################################
+        //! The CPU queue interface
+        template<
+            typename TDev>
+        class IGenericQueue
```

Comment:
I really do not like the `Generic` names for the queues and the event here. They are not that generic at all. The `Cpu` part of the name previously had two meanings:
* the queue holds `EventCpu`
* the queue is executed on the CPU

You made the first point generic but what it still differentiates from a generic queue or the CUDA queues is that it can only be run on the host CPU.
I would prefer if this would be called `IGenericCpuQueue` and still be located in the `cpu` subfolder. The same is true for the queue impls and the events.

Reply:
OK, I thought the name was only related to the Dev / Event type.

Howver, the name `CPU` then is still too general in that it only says that it lives on the host. There are multiple ways to implement a device-wise generic queue. This on used C++11 threads, on different option would be OpenMP tasks. With this in mind, we could call them, e.g. `IGenericThreadsQueue`, `QueueThreadsBlocking`. The interface would still go into the cpu subfolder.

Reply:
I like your proposed names

Reply:
I meant to write `QueueGenericThreadsBlocking` instead of `QueueThreadsBlocking`, Is that ok, too?

Reply:
Also fine